server/request.go

```diff
@@ -107,75 +108,99 @@ func (s *Server) GetPhotoIDs(w http.ResponseWriter, r *http.Request) {
 		Success: true,
 		IDs:     ids,
 	}
-	WriteJsonResponse(v, w)
+	WriteJsonResponse(v, 200, w)
 }
 
 func (s *Server) PostPhoto(w http.ResponseWriter, r *http.Request) {
-	r.ParseMultipartForm(1 << 23) // 8M max memory
+	r.ParseMultipartForm(1 << 25) // 32M max memory
 
 	input, _, err := r.FormFile("photo")
 	if err != nil {
 		log.Info(err)
-		WriteError("Unable to parse form value 'photo'", 400, w)
+		WriteError("unable to parse form value 'photo'", 400, w)
 		return
 	}
 	defer input.Close()
 
 	id, err := GenPhotoID(input)
 	if err != nil {
 		log.Error(err)
-		WriteError("Unable to generate photo ID", 500, w)
+		WriteError("unable to generate photo ID", 500, w)
 		return
 	}
 
-	var existingPhoto Photo
-	if err := s.db.One("ID", id, &existingPhoto); err == nil {
-		log.Info(fmt.Sprintf("Photo %s already exists", id))
-		WriteError("Photo already exists", 400, w)
-		return
+	overwrite := r.FormValue("overwrite")
+	photo, err := s.GetPhotoFromDatabase(id)
+	if err == nil {
+		if overwrite == "true" {
+			s.db.DeleteStruct(photo)
+		} else {
+			WriteError("photo already exists", 400, w)
+			return
+		}
 	} else if err != storm.ErrNotFound {
 		log.Error(err)
-		WriteError("Unable to query database", 500, w)
+		WriteError("unable to read from database", 500, w)
 		return
 	}
 
-	photoPath := path.Join(s.cfg.ImgFolder(), fmt.Sprintf("%s.jpg", id))
-	thumbPath := path.Join(s.cfg.ImgFolder(), fmt.Sprintf("%s-thumb.jpg", id))
-
-	xif, rect, err := ProcessPhoto(input, id, photoPath, thumbPath)
-	if err != nil {
-		log.Error("Failed to save image")
-		WriteError("Unable to save image", 500, w)
-
-		os.Remove(photoPath)
-		os.Remove(thumbPath)
-		return
+	now := time.Now()
+	photo = Photo{
+		ID:              id,
+		UploadedAt:      &now,
+		Status:          Processing,
+		StatusUpdatedAt: &now,
 	}
 
-	newPhoto := NewPhoto(id, *rect, *xif)
-
-	if err := s.db.Save(newPhoto); err != nil {
+	if err := s.db.Save(&photo); err != nil {
 		log.Error(err)
-		WriteError("Unable to save to database", 500, w)
-
-		os.Remove(photoPath)
-		os.Remove(thumbPath)
+		WriteError("unable to write to database", 500, w)
 		return
 	}
-
 	v := PostPhotoResponse{
 		Success: true,
 		NewID:   id,
+		Status:  Processing,
 	}
-	WriteJsonResponse(v, w)
+	WriteJsonResponse(v, 200, w)
+
+	go func() {
```

Comment:
If the server stops while it is processing a photo, it never gets processed, and any further attempts to re-upload the photo return an "already exists" error. It looks like this means we end up with orphaned .jpg and -thumb.jpg files in the data folder.

We could fix this by looking for photos with status "processing" on server startup and attempting to process them before starting the HTTP server. It could also scan for and remove .jpg files that aren't referenced by any entries in the database.

Reply:
If you specify the overwrite flag, then you can reupload the same photograph

Reply:
also the jpgs get cleaned up if they don't process correctly, it's a failsafe in the PostPhoto function around L180 of request.go